ibc-clients/cw-context/src/context/mod.rs

```diff
@@ -0,0 +1,292 @@
+                    let processed_height_key = self.client_update_height_key(&height);
+                    metadata.push(GenesisMetadata {
+                        key: processed_height_key.clone(),
+                        value: self.retrieve(&processed_height_key)?,
+                    });
+                    let processed_time_key = self.client_update_time_key(&height);
+                    metadata.push(GenesisMetadata {
+                        key: processed_time_key.clone(),
+                        value: self.retrieve(&processed_time_key)?,
+                    });
+                }
+                None => continue,
+            }
+        }
+
+        let iterator = self
+            .storage_ref()
+            .range(Some(&start_key), None, Order::Ascending);
+
+        for (key, height) in iterator {
+            metadata.push(GenesisMetadata { key, value: height });
+        }
+
+        Ok(Some(metadata))
+    }
+
+    /// Returns the checksum of the current contract.
+    pub fn obtain_checksum(&self) -> Result<Checksum, ClientError> {
+        match &self.checksum {
+            Some(checksum) => Ok(checksum.clone()),
+            None => {
+                let client_state_value = self.retrieve(ClientStatePath::leaf())?;
+
+                let wasm_client_state: WasmClientState =
+                    Protobuf::<Any>::decode(client_state_value.as_slice()).map_err(|e| {
+                        ClientError::Other {
+                            description: e.to_string(),
+                        }
+                    })?;
+
+                Ok(wasm_client_state.checksum)
+            }
+        }
+    }
+
+    /// Encodes the given client state into a byte vector.
+    pub fn encode_client_state(
+        &self,
+        client_state: C::ClientState,
+    ) -> Result<Vec<u8>, ClientError> {
+        let wasm_client_state = WasmClientState {
+            data: C::ClientState::encode_thru_any(client_state.clone()),
+            checksum: self.obtain_checksum()?,
+            latest_height: client_state.latest_height(),
+        };
+
+        Ok(Any::from(wasm_client_state).encode_to_vec())
+    }
+}
+
+pub trait StorageRef {
+    fn storage_ref(&self) -> &dyn Storage;
+}
+
+impl<'a, C: ClientType<'a>> StorageRef for Context<'a, C> {
+    fn storage_ref(&self) -> &dyn Storage {
+        match self.deps {
+            Some(ref deps) => deps.storage,
+            None => match self.deps_mut {
+                Some(ref deps) => deps.storage,
+                None => panic!("Either deps or deps_mut should be available"),
+            },
+        }
+    }
+}
+
+pub trait StorageMut: StorageRef {
+    fn storage_mut(&mut self) -> &mut dyn Storage;
+}
+
+impl<'a, C: ClientType<'a>> StorageMut for Context<'a, C> {
+    fn storage_mut(&mut self) -> &mut dyn Storage {
+        match self.deps_mut {
+            Some(ref mut deps) => deps.storage,
+            None => panic!("deps_mut should be available"),
+        }
```

Comment:
You should be able to do following:

```rs
self.deps_mut.as_ref_mut().map(|deps| deps.storage).expect("deps_mut should be available");
```

Reply:
https://github.com/cosmos/ibc-rs/pull/1167#discussion_r1569178327